drivers/i2s/i2s_ll_stm32.c

```diff
@@ -597,12 +639,41 @@ static void dma_tx_callback(const struct device *dma_dev, void *arg,
 	k_mem_slab_free(stream->cfg.mem_slab, stream->mem_block);
 	stream->mem_block = NULL;
 
+	/* Mark the tx sample completion */
+	stream->sample_tx_block_dma_sent++;
+
 	/* Stop transmission if there was an error */
 	if (stream->state == I2S_STATE_ERROR) {
 		LOG_ERR("TX error detected");
 		goto tx_disable;
 	}
 
+	/*
+	 * Check if the tx stop command can be handled
+	 * if True, the tx stop command can be handled
+	 */
+	if ((stream->state == I2S_STATE_STOPPING) && (status == DMA_STATUS_COMPLETE)) {
+		/*
+		 * Check if all tx samples have been completely handled
+		 * as stated in zephyr i2s specification, in case of DRAIN command
+		 * send all data in the transmit queue and stop the transmission.
+		 */
+		if (stream->sample_tx_block_sent == stream->sample_tx_block_dma_sent) {
```

Comment:
Nit: what if you replace this with `queue_is_empty`?

Reply:
I replace it with queue_is_empty , it didn't solve the last block problem but its more logical
 because the variable stream->sample_tx_block_sent it is initialized to 0 but never seems to be updated .